### PR TITLE
Add support for more advanced `SlugInput` overrides with `locale` & `formatters`

### PR DESCRIPTION
**Pull request by lb-**
> [!IMPORTANT]  
> Builds on PR #12554 - Which does not change main functionality but instead makes the code easier to maintain/test.

## Description

This PR primarily introduces two new super powers for our Django `SlugInput` widget (which uses the Stimulus `CleanController`).

1. `SlugInput` transliteration is now locale aware and there is a documented way to have finer grained control over how this works.
2. `SlugInput` now supports `formatters`, a list of regex/substitute entries, covering a lot of common use cases when custom slug generation is required.

### Related PR

As noted above, this PR builds on another PR (https://github.com/wagtail/wagtail/pull/12554) that makes the base JS `CloneController` better unit tested, fixes a few small edge cases inconsistencies and adds read only event dispatching. In addition, it should be easier to maintain & easier to override with JS class extension if the `SlugInput` super powers aren't enough.

### Locale

The `locale` aware urlify in `CleanController` makes some assumptions about the target locale based on the `ACTIVE_CONTENT_LOCALE` which is available in JavaScript when i18n is enabled and there is a target locale for the page. For most languages this will make no difference, but for the overlapping Russian,  Belarusian & Ukrainian characters, this will now reflect the correct transliteration.

Developers can also override this by easily opting out of this behavior by passing different `locale` values to the `SlugInput`.

### Formatters

The `format` action method in the `CleanController` adds support for and array of regex pairs (a pattern & replace string, or pattern/flags & replace string). The `format` method will run whenever slugify/urlify is used, but can also be run directly for non-`SlugInput` usage int he future, this should make https://github.com/wagtail/wagtail/issues/7417 pretty much done (on the JavaScript side anyway).

The `SlugInput` will also accept a `formatters` list, this will be a similar structure but better suited to the way that Python regex's work. There's a new util that will convert the most common cases of Python regex to the equivalent JavaScript regex. There's nuance here, and instead of solving for everything, I have tried to get a solid start and we can enhance over time if there are new requests, most importantly there's solid error handling in both the Python conversion and the regex parsing in the Stimulus controller.

### Documentation

I have added a new section in the ['Advanced > Customizing Wagtail > Customizing the editing interface'](https://wagtail--12598.org.readthedocs.build/en/12598/advanced_topics/customization/page_editing_interface.html#customizing-page-slug-generation) page, specifically about customizing the slug generation.

This first talks to the different ways to override the `SlugInput` and then talks to the locale/formatters usage with a suite of examples for each.

I have added some considerations for these also so that developers know what to expect when changing behavior here, the way the title/slug sync works can be a bit non-intuitive at first.

I'm not sure if I should instead put the new docs on this page https://docs.wagtail.org/en/stable/extending/forms.html

### Other changes of note

- `urlify.config.json` now supports language codes defined on the config, I opted to keep this structure mostly the same.
- Added better TypeScript typing for the global `WagtailConfig`.
- The page copy form, because it knows the target locale, will get passed in this locale. However, this is not really used as there's no title sync to the slug input on this page (it just uses slugify, not urlify, urlify is the one that does unicode handling). I decided to do this for consistency.
- I have brought in the `castArray` util (exact commit), from a separate PR https://github.com/wagtail/wagtail/pull/11202

## Screenshots

For the example below, I have bundled a bunch of items together into one test. In the code snippet you can see we are using 3 formatters.

Hence, the formatted transliteration should be this;

```js
"12 нехай веселощі the починаються";
"blog-n-nehaj-veseloshi-pochinayutsya"; // Russian (ur lang & reflects default transliteration)
"blog-n-nekhai-veseloshchi-pochynayutsya"; // Ukrainian (uk lang)
// Observe the different letter replacement, but both add 'blog-' and both replace the 12 with the letter 'n'
```

```py
class StandardPageForm(WagtailAdminPageForm):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields["slug"].widget = SlugInput(
            formatters = [
                r'(?i)\b(?:and|or|the|in|of|to)\b', # remove common stop words
                [r'\d+', 'n'], # replace numbers with 'n'
                [re.compile(r"^(?!blog[-\s])", flags=re.MULTILINE), 'blog-'], # add 'blog-' prefix if not already present
            ]
        )

class StandardPage(Page):
    base_form_class = StandardPageForm
    ## ...
```

| Ukrainian | Russian | i18n off (en) |
| --------- | ------- | ------------- |
| i18n-uk  | i18n-ru |   i18n-off-en-user  |

### Prerequisites

- `ALLOW_UNICODE_SLUGS` needs to be `False` to see any unicode to ascii conversion.
- Title to slug sync will NOT happen for published pages, only for the new page form or pages that are unpublished.
- Title to slug sync will NOT happen if the values are determined to be out of sync when focusing on the title. This is intentional, if changing the `SlugInput` items, you will need to go to the promote tab, clear the slug field and then edit the title, this will update the slug and subsequent changes to title can be correctly validated.
- It's best to do a mix of testing with i18n enabled/disabled and different user languages (if i18n is disabled).

### Scenarios to test

1. New page creation - this is the most common usage of slug generation.
2. Editing an existing page - when unpublished.
3. Copying a page and editing the slug, then blurring (as noted above, should be no functional changes here).
4. Storybook page for `CleanController`.

### Browsers tested

Tested on the following browsers.

- Firefox 132
- Safari 17.5
- Chrome 131

---

## Proposed release notes

- Fix: Refine Russian and add  Ukrainian + Belarusian transliteration support when resolving ASCII slugs
- Feature: Add support for `locale` overrides in the `SlugInput` widget
- Feature: Add support for `formatters` in the `SlugInput` widget, allowing more flexible slug generation
- Documentation: Add new [Customizing `SlugInput` widget](customizing_slug_widget) documentation

### Changes to `SlugInput` transliteration impacting Russian/Ukrainian/Belarusian Cyrillic characters

For most transliteration, there should be no difference to the previous release. This release introduces `locale` ware slug transliteration for overlapping characters across different languages.

This means that existing slugs may be out of sync with their titles, assuming your editors have not manually adjusted the slug value, subsequent changes to titles on unpublished pages may not further sync up the slug field.

To work around this, you can advise users to manually clear the slug fields, then focus on the title field to get the new slug generated. Being aware of impacts to existing URLs if a page was previously published.

Reminder that this only impacts unpublished, existing, pages.

The default `SlugInput` usage of this locale has some nuances to be aware of.

If your project **uses i18n** (internationalization, `WAGTAIL_I18N_ENABLED = True`), has set `WAGTAIL_ALLOW_UNICODE_SLUGS = False` and has Ukrainian or Belarusian targeted pages, the slug generation will now be correctly locale aware and have better transliteration for overlapping characters.

If your project does **not use i18n** (internationalization, `WAGTAIL_I18N_ENABLED = False` or not set), but still has set `WAGTAIL_ALLOW_UNICODE_SLUGS = False` **and** you want to target a specific locale for all your slugs (or on a sub-set of page models).

It's now possible to force a specific locale for the `SlugInput` widget, either globally or for specific Page models, it's also possible to opt out completely of the above translation focused behavior.

See [](customizing_slug_widget) for now to do this.

### CODE REVIEW

File: client/src/controllers/CleanController.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    const action = Actions.Identity;
+    const value = this.element.value;
+    this.applyUpdate(action, value, value);
+    return value;
+  }
+
+  /**
+   * Formats the source value based on supplied formatters and formatting options.
+   * Runs as part of the prepare method on methods called (when used as Stimulus actions)
+   * or can be called directly.
```
Comment (lb-) [inline client/src/controllers/CleanController.ts:139]:
```suggestion
   * or can be used directly as a standalone action.
```

File: client/src/utils/urlify.config.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [
   {
-    "LATIN_MAP": [
+    "LATIN_MAP:foo": [
```
Comment (lb-) [inline client/src/utils/urlify.config.json:3]:
```suggestion
    "LATIN_MAP": [
```

Fix-up needed.

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 3]
```diff
@@ -292,3 +292,166 @@ class BlogPage(Page):
...
+    ]
+```
+
+### Overriding `SlugInput`
+
+There are multiple ways to override the `SlugInput`, depending on your use case and admin setup.
+
+#### Via `promote_panels`
+
+The simplest, if have already set custom `promote_panels` is to leverage the `FieldPanel` widget kwarg as follows.
```
Comment (lb-) [inline docs/advanced_topics/customization/page_editing_interface.md:320]:
```suggestion
The simplest, if have already set custom `promote_panels`, is to leverage the `FieldPanel` widget kwarg as follows.
```

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 4]
```diff
@@ -292,3 +292,166 @@ class BlogPage(Page):
...
+    # no need to declare `promote_panels`
+```
+
+#### Globally, via `register_form_field_override`
+
+If you want to override the `models.SlugField` widget for the entire admin, this can be done using the Wagtail admin util.
+
+It's best to add this to a `wagtail_hook.py` file as this will get run at the right time.
+
+```py
```
Comment (lb-) [inline docs/advanced_topics/customization/page_editing_interface.md:364]:
```suggestion
```py
# wagtail_hooks.py
```

Add other file names in also.

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 5]
```diff
@@ -292,3 +292,166 @@ class BlogPage(Page):
...
+    models.SlugField,
+    override={"widget": widgets.SlugInput(locale="uk-UK")},
+)
+```
+
+The following sections will only focus on the `SlugInput(...` usage and not where to override this.
+
+### Overriding the default locale behavior via `locale`
+
+The `SlugInput` is locale aware and will adjust the transliteration (Unicode to ASCII conversion) based on the determined model locale, only when [`ALLOW_UNICODE_SLUGS`](wagtail_allow_unicode_slugs) is `False`.
```
Comment (lb-) [inline docs/advanced_topics/customization/page_editing_interface.md:382]:
```suggestion
The `SlugInput` is locale aware and will adjust the transliteration (Unicode to ASCII conversion) based on the most suitable locale, only when [`ALLOW_UNICODE_SLUGS`](wagtail_allow_unicode_slugs) is `False`.
```

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 6]
```diff
@@ -292,3 +292,166 @@ class BlogPage(Page):
...
+
+The locale will be determined from the target translation locale if ()[internationalisation] is enabled, or the current admin user's language if not.
```
Comment (lb-) [inline docs/advanced_topics/customization/page_editing_interface.md:384]:
```suggestion
The locale will be determined from the target translation locale if ()[internationalisation] is enabled. If Internationalization is not in use, it will be based on the language of the admin for the current logged in user. This can be overridden, see below.
```

File: wagtail/admin/widgets/slug.py [hunk 7]
```diff
@@ -1,24 +1,78 @@
...
+            # If a locale is provided as a string, return it, otherwise assume it's a Locale instance
+            # If no locale can be found,  'und' - as per ISO639-2 standard, 'und' represents 'undetermined'
+            # If the attribute is not set, the input will fall back to the ACTIVE_LOCALE or document lang.
+            undetermined_locale = "und"
+            locale = (
+                (locale or undetermined_locale)
+                if isinstance(locale, str)
+                else getattr(locale, "language_code", undetermined_locale)
+            )
+            default_attrs["data-w-slug-locale-value"] = locale
```
Comment (lb-) [inline wagtail/admin/widgets/slug.py:73]: I can probably simplify this. 

```py
            locale = getattr(locale, "language_code", locale) or undetermined_locale
```

I'll need to update the description also.

File: client/src/utils/urlify.config.json [hunk 8]
```diff
@@ -175,7 +175,7 @@
     ]
   },
   {
-    "RUSSIAN_MAP": [
+    "RUSSIAN_MAP:ru": [
```
Comment (ACK1D) [inline client/src/utils/urlify.config.json:178]: ["х", "h"], -> ["х", "x"] or  ["х", "kh"]
["Х", "H"], -> ["Х", "X"] or  ["X", "Kh"]
["Щ", "Sh"], -> ["Щ", "Shh"]
["щ", "sh"], -> ["щ", "shh"]

see here: https://ifap.ru/library/gost/70342014.pdf (page 7 in pdf)

I would also reconsider transliteration for the Belarusian language, if possible

**Comment by lb-** [comment]
OK, this is now ready for review.

File: client/src/utils/urlify.test.js [hunk 9]
```diff
@@ -9,6 +9,36 @@ describe('urlify', () => {
...
+      expect(urlify('  I like _ßpaces')).toBe('-i-like-_sspaces');
+    });
+
+    it('should allow a supplied locale to re-order downcode mapping', () => {
+      // mix of characters that are transliterated differently across Russian/Ukrainian
+      const chars = 'Ц-йи-Ґ-Х-щ';
+      const ruTransliteration = 'c-ji-g-h-sh';
+      const ukTransliteration = 'ts-iy-g-kh-shch';
+      expect(urlify(chars)).toEqual(ruTransliteration); // by default the Russian values come through with priority
+      expect(urlify('Ц-йи-Ґ-Х-щ', { locale: 'ru' })).toEqual(ruTransliteration);
```
Comment (ACK1D) [inline client/src/utils/urlify.test.js:23]: There is no letter “Ґ” in the Russian language. Ukrainian has “Ґ” and “Г”, Russian has only “Г”

File: client/src/utils/urlify.config.json (hunk 8 above)
Comment (lb-) [reply client/src/utils/urlify.config.json:178]: @ACK1D I assume this would be a separate language code?

Is this correct?

```json
  # after Ukrainian
  {
    "BELARUSIAN_MAP:bu": [
      ["X", "Kh"],
      ["х", "kh"],
      ["Х", "X"],
      ["х", "x"],
      ["щ", "shh"],
      ["Щ", "Shh"]
    ]
  },
```

Note: We probably would add this as a separate PR, but just want to get the config right.

File: client/src/utils/urlify.test.js (hunk 9 above)
Comment (lb-) [reply client/src/utils/urlify.test.js:23]: I understand that this gibberish set of strings may not be used, the idea here is to smoke test a mix of characters across conflicting character sets.

Hopefully that's ok as is.

File: client/src/utils/urlify.config.json (hunk 8 above)
Comment (ACK1D) [reply client/src/utils/urlify.config.json:178]: @lb- no, what I sent is the incorrect current implementation for Russian according to Russian GOSTs. But, if we have disturbed this hive with the difference between Russian and Ukrainian, then it makes sense to make corrections for the Belarusian language as well.  

`{
    "RUSSIAN_MAP:ru": [
      ["а", "a"],
      ["б", "b"],
      ["в", "v"],
      ["г", "g"],
      ["д", "d"],
      ["е", "e"],
      ["ё", "yo"],
      ["ж", "zh"],
      ["з", "z"],
      ["и", "i"],
      ["й", "j"],
      ["к", "k"],
      ["л", "l"],
      ["м", "m"],
      ["н", "n"],
      ["о", "o"],
      ["п", "p"],
      ["р", "r"],
      ["с", "s"],
      ["т", "t"],
      ["у", "u"],
      ["ф", "f"],
      ["х", "x"], // current: "х", "h" but aslo allowed: "х", "kh"
      ["ц", "c"],
      ["ч", "ch"],
      ["ш", "sh"],
      ["щ", "shh"], // current: "щ", "sh"
      ["ъ", ""],
      ["ы", "y"],
      ["ь", ""],
      ["э", "e"],
      ["ю", "yu"],
      ["я", "ya"],
      ["А", "A"],
      ["Б", "B"],
      ["В", "V"],
      ["Г", "G"],
      ["Д", "D"],
      ["Е", "E"],
      ["Ё", "Yo"],
      ["Ж", "Zh"],
      ["З", "Z"],
      ["И", "I"],
      ["Й", "J"],
      ["К", "K"],
      ["Л", "L"],
      ["М", "M"],
      ["Н", "N"],
      ["О", "O"],
      ["П", "P"],
      ["Р", "R"],
      ["С", "S"],
      ["Т", "T"],
      ["У", "U"],
      ["Ф", "F"],
      ["Х", "X"], // current: "Х", "H" (changed to X, but aslo allowed: "Х", "Kh")
      ["Ц", "C"],
      ["Ч", "Ch"],
      ["Ш", "Sh"],
      ["Щ", "Shh"], // current: "Щ", "Sh" (changed to Shh)
      ["Ъ", ""],
      ["Ы", "Y"],
      ["Ь", ""],
      ["Э", "E"],
      ["Ю", "Yu"],
      ["Я", "Ya"]
    ]
  },`

Comment (lb-) [reply client/src/utils/urlify.config.json:178]: Ahh I see. I had a feeling that maybe the Russian ones were incorrect, it's from a port of the Django mapping. It's probably incorrect there.

I'll take a look, it might be best to do as you suggested and fix up these also in the same PR. 

Out of curiosity, because that PDF is a bit hard to parse for me. 

Does this site reflect the desired behaviour?
https://translit.cc/

Or do you know of something equivalent to this website?
https://czo.gov.ua/en/translit

Or is that PDF the best source of truth. I'll probably add these links to the JS code also.

Comment (ACK1D) [reply client/src/utils/urlify.config.json:178]: if you are select GOST 7.79 B in https://translit.cc/ then my proposed changes are relevant and correct. You can see it here: https://en.wikipedia.org/wiki/GOST_7.79-2000 

I don't know of any https://czo.gov.ua/en/translit analogs, but I'll ask around.

Comment (ACK1D) [reply client/src/utils/urlify.config.json:178]: if we talk about the Belarusian “default”:
```
{
    "BELARUSIAN_MAP:be": [
      ["а", "a"],
      ["б", "b"],
      ["в", "v"],
      ["г", "g"],
      ["д", "d"],
      ["е", "je"], // different to russian
      ["ё", "jo"], // different to russian
      ["ж", "zh"],
      ["з", "z"],
      ["і", "i"], // different to russian letter
      ["й", "j"],
      ["к", "k"],
      ["л", "l"],
      ["м", "m"],
      ["н", "n"],
      ["о", "o"],
      ["п", "p"],
      ["р", "r"],
      ["с", "s"],
      ["т", "t"],
      ["у", "u"],
      ["ў", "w"], // new letter
      ["ф", "f"],
      ["х", "h"], // if we are change ru to ["х", "x"] or ["х", "kh"]
      ["ц", "c"],
      ["ч", "ch"],
      ["ш", "sh"],
      ["ы", "y"],
      ["ь", ""],
      ["э", "e"],
      ["ю", "ju"], // different to russian
      ["я", "ja"], // different to russian
      ["А", "A"],
      ["Б", "B"],
      ["В", "V"],
      ["Г", "G"],
      ["Д", "D"],
      ["Е", "Je"], // different to russian
      ["Ё", "Jo"], // different to russian
      ["Ж", "Zh"],
      ["З", "Z"],
      ["І", "I"], // different to russian letter
      ["Й", "J"],
      ["К", "K"],
      ["Л", "L"],
      ["М", "M"],
      ["Н", "N"],
      ["О", "O"],
      ["П", "P"],
      ["Р", "R"],
      ["С", "S"],
      ["Т", "T"],
      ["У", "U"],
      ["Ў", "W"], // new letter
      ["Ф", "F"],
      ["Х", "H"], // if we are change ru to ["Х", "X"] or ["Х", "Kh"]
      ["Ц", "C"],
      ["Ч", "Ch"],
      ["Ш", "Sh"],
      ["Ъ", ""],
      ["Ы", "Y"],
      ["Ь", ""],
      ["Э", "E"],
      ["Ю", "Ju"], // different to russian
      ["Я", "Ja"] // different to russian
    ]
  },
  ```

**Comment by lb-** [comment]
Thanks for the inputs @ACK1D  - please take another look if you have time.

In addition, I have removed the fallback to reading the `document.lang`, this felt like the wrong direction and assumes too much about editor's setups. I will refine the proposed release notes / upgrade considerations in the description shortly.

Note, I have a backup of the previous branch here https://github.com/lb-/wagtail/tree/feature/11916-enhanced-slug-input-customisation-backup

File: client/src/utils/urlify.config.json [hunk 10]
```diff
@@ -245,19 +245,100 @@
...
+      ["С", "S"],
+      ["Т", "T"],
+      ["У", "U"],
+      ["Ў", "W"],
+      ["Ф", "F"],
+      ["Х", "H"],
+      ["Ц", "C"],
+      ["Ч", "Ch"],
+      ["Ш", "Sh"],
+      ["Ъ", ""],
```
Comment (ACK1D) [inline client/src/utils/urlify.config.json:332]: That's redundant (["Ъ", ""],). I removed the lowercase letter when I adapted the Russian version, but forgot about the uppercase letter.

File: client/src/utils/urlify.test.js [hunk 11]
```diff
@@ -9,6 +9,36 @@ describe('urlify', () => {
...
+
+      // another example
+      expect(urlify('Георгій', { locale: 'uk-UK' })).toEqual('heorhii');
+      expect(urlify('Георгій', { locale: 'en' })).toEqual('georgij');
+      expect(urlify('Георгій')).toEqual('georgij');
+      expect(urlify('Георгій', { locale: 'ru' })).toEqual('georgij');
+    });
+
+    it('should support locale with normal urlify whitespace handling', () => {
+      const title = 'жйи Проєкт   Герой Їжак Георгій ';
```
Comment (ACK1D) [inline client/src/utils/urlify.test.js:34]: I don't know if it's important, but there's no meaningfulness here. It's just a bunch of letters

File: client/src/utils/urlify.test.js [hunk 12]
```diff
@@ -9,6 +9,36 @@ describe('urlify', () => {
...
+      // mix of characters that are transliterated differently across Russian/Ukrainian
+      const chars = 'Ц-йи-Ґ-Х-щ';
+      const ruTransliteration = 'c-ji-g-x-shh';
+      const ukTransliteration = 'ts-iy-g-kh-shch';
+      expect(urlify(chars)).toEqual(ruTransliteration); // by default the Russian values come through with priority
+      expect(urlify('Ц-йи-Ґ-Х-щ', { locale: 'ru' })).toEqual(ruTransliteration);
+      expect(urlify('Ц-йи-Ґ-Х-щ', { locale: 'uk' })).toEqual(ukTransliteration);
+
+      // another example
+      expect(urlify('Георгій', { locale: 'uk-UK' })).toEqual('heorhii');
```
Comment (ACK1D) [inline client/src/utils/urlify.test.js:27]: In my opinion it is not quite right to take a word in one language and see how they differ in transliteration with different localization. In this case I would use another name (not Георгий(ru)/Георгій(uk) but Богдан(ru/uk)), because it is spelled the same in both languages and is not misleading.

Comment (lb-) [reply client/src/utils/urlify.test.js:27]: Great. I'll update this unit test. Thank you.

File: client/src/utils/urlify.test.js (hunk 11 above)
Comment (lb-) [reply client/src/utils/urlify.test.js:34]: Correct, it's just an example to test the differences. 

In my mind 'real world' examples are nice to have but the primary goal is to test the differences and validation of expected behaviour.

File: client/src/utils/urlify.test.js (hunk 12 above)
Comment (lb-) [reply client/src/utils/urlify.test.js:27]: @ACK1D any chance you have an example I can use that's 'real world ish' for the Russian/ Belarusian difference. I'd like to add a unit test for that also.

File: client/src/utils/urlify.config.json (hunk 10 above)
Comment (lb-) [reply client/src/utils/urlify.config.json:332]: Great. I'll fix this up. 

Thank you so much for your help here!

File: client/src/utils/urlify.test.js (hunk 12 above)
Comment (ACK1D) [reply client/src/utils/urlify.test.js:27]: > @ACK1D any chance you have an example I can use that's 'real world ish' for the Russian/ Belarusian difference. I'd like to add a unit test for that also.

As an option to check the Russian sentence for transliteration in the Belarusian localization (in theory it should break because of the absence of the letters “И” and “Щ” in the language, I think for these cases it is necessary to take the Russian localization. I mean in priority the Belarusian localization, but if some letters are missing, then still look at the Russian version, because in Belarus there are two official languages Russian and Belarusian).
Sentence: "Бен делает отличные вещи для сообщества трясогузки!"

Comment (lb-) [reply client/src/utils/urlify.test.js:27]: ohh very kind, I have refined it a bit but will use this. Thank you.

**Comment by lb-** [comment]
Rebased

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 13]
```diff
@@ -292,3 +292,176 @@ class BlogPage(Page):
...
+
+### Adding custom formatters via `formatters`
+
+`SlugInput` also accepts a `formatters` kwarg, allowing a list of custom formatters to be applied, each formatter item can be one of the following.
+
+1. A regex pattern or string (for example `r"\d"`, `re.compile(r"\d")`)
+2. A list-like value containing a regex pattern/string and a replacement string (for example `[r"\d", "n"]`)
+3. A list-like value containing a regex pattern/string and a replacement string & custom base JavaScript regex flags (for example `[r"\d", "n", "u"]`)
+
+A reminder to ensure that regex strings are appropriately escaped.
```
Comment (zerolab) [inline docs/advanced_topics/customization/page_editing_interface.md:410]:
```suggestion
As a reminder, ensure that regex strings are appropriately escaped.
```

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 14]
```diff
@@ -292,3 +292,176 @@ class BlogPage(Page):
...
+
+As an example below, here's a formatter that will remove common stop words from the slug. If the title is entered as `The weather and the times`, this will produce a slug of `weather-times`.
```
Comment (zerolab) [inline docs/advanced_topics/customization/page_editing_interface.md:412]:
```suggestion
For example, here's a formatter that will remove common stop words from the slug. If the title is entered as `The weather and the times`, this will produce a slug of `weather-times`.
```

**Review by zerolab — COMMENTED** [review]
Thank you for this @lb-  (and sorry for the rather delayed local testing).

Added a couple of suggestions for tidying up the docs.

Tested locally by changing the `SlugInput` locale via `register_form_field_override` and it works well.

The JS changes make sense to me - would still be good to get someone with stronger JS chops to double check it.

**Comment by lb-** [comment]
Thanks @zerolab for trying it out locally, glad it all worked well. 

I really appreciate the docs improvements and will get those cleaned up soon + a rebase.

**Comment by lb-** [comment]
All the docs recommended fixes have been applied. Rebased.

**Comment by zerolab** [comment]
@ACK1D any chance you could give this one final check?

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 15]
```diff
@@ -340,3 +340,176 @@ class BlogPage(Page):
...
+    models.SlugField,
+    override={"widget": widgets.SlugInput(locale="uk-UK")},
+)
+```
+
+The following sections will only focus on the `SlugInput(...)` usage and not where to override this.
+
+### Overriding the default locale behavior via `locale`
+
+The `SlugInput` is locale aware and will adjust the transliteration (Unicode to ASCII conversion) based on the most suitable locale, only when [`ALLOW_UNICODE_SLUGS`](wagtail_allow_unicode_slugs) is `False`.
```
Comment (gasman) [inline docs/advanced_topics/customization/page_editing_interface.md]:
```suggestion
The `SlugInput` is locale aware and will adjust the transliteration (Unicode to ASCII conversion) based on the most suitable locale, only when [`WAGTAIL_ALLOW_UNICODE_SLUGS`](wagtail_allow_unicode_slugs) is `False`.
```

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 16]
```diff
@@ -340,3 +340,176 @@ class BlogPage(Page):
...
+
+The locale will be determined from the target translation locale if ()[internationalisation] is enabled.
```
Comment (gasman) [inline docs/advanced_topics/customization/page_editing_interface.md]:
```suggestion
The locale will be determined from the target translation locale if [](internationalisation) is enabled.
```

File: docs/advanced_topics/customization/page_editing_interface.md [hunk 17]
```diff
@@ -340,3 +340,176 @@ class BlogPage(Page):
...
+    ]
+```
+
+### Overriding `SlugInput`
+
+There are multiple ways to override the `SlugInput`, depending on your use case and admin setup.
+
+#### Via `promote_panels`
+
+The simplest, if have already set custom `promote_panels`, is to leverage the `FieldPanel` widget kwarg as follows.
```
Comment (gasman) [inline docs/advanced_topics/customization/page_editing_interface.md]:
```suggestion
The simplest, if you have already set custom `promote_panels`, is to leverage the `FieldPanel` widget kwarg as follows.
```

**Review by gasman — APPROVED** [review]
Great work @lb-! Very happy with the way that it does the right thing with transliterations out of the box, but also provides more extensive customisability where needed.

Just a couple of documentation fixes, which I'll apply while merging.

**Comment by gasman** [comment]
Merged in c1ccffbe142d0216ba175e0677b85381e52b71c8

File: docs/advanced_topics/customization/page_editing_interface.md (hunk 16 above)
Comment (lb-) [reply docs/advanced_topics/customization/page_editing_interface.md]: Thanks for this, it's interesting the CI didn't flag this bad link.

**Comment by lb-** [comment]
Thank you @gasman & @zerolab for the reviews, feedback and suggestions on different approaches. 

Glad to see this in.
